src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,392 @@
+    default:
+        m_isValid = false;
+        DEBUG_ASSERT(!"Unsupported format");
+    }
+
+    if (!m_isValid) {
+        return;
+    }
+
+    prepare();
+}
+
+void ControllerRenderingEngine::prepare() {
+    m_pThread = std::make_unique<QThread>();
+    m_pThread->setObjectName("ControllerScreenRenderer");
+
+#if QT_VERSION >= QT_VERSION_CHECK(6, 7, 0)
+    [[maybe_unused]] bool successful = moveToThread(m_pThread.get());
+    DEBUG_ASSERT(successful);
+#else
+    moveToThread(m_pThread.get());
+#endif
+
+    // these at first sight weird-looking connections facilitate thread-safe communication.
+    connect(this,
+            &ControllerRenderingEngine::engineSetupRequested,
+            this,
+            &ControllerRenderingEngine::setup);
+    connect(this,
+            &ControllerRenderingEngine::sendFrameDataRequested,
+            this,
+            &ControllerRenderingEngine::send);
+    connect(m_pThread.get(),
+            &QThread::finished,
+            this,
+            &ControllerRenderingEngine::finish);
+
+    m_pThread->start(QThread::NormalPriority);
+}
+
+ControllerRenderingEngine::~ControllerRenderingEngine() {
+    DEBUG_ASSERT(QThread::currentThread() != thread());
+    m_pThread->wait();
+    VERIFY_OR_DEBUG_ASSERT(!m_fbo) {
+        kLogger.critical() << "The ControllerEngine is being deleted but hasn't been "
+                              "cleaned up. Brace for impact";
+    };
+}
+
+void ControllerRenderingEngine::start() {
+    VERIFY_OR_DEBUG_ASSERT(!thread()->isFinished() && !thread()->isInterruptionRequested()) {
+        kLogger.critical() << "Render thread has or is about to terminate. Cannot "
+                              "start this render anymore.";
+        return;
+    }
+    QCoreApplication::postEvent(this, new QEvent(QEvent::UpdateRequest));
+}
+bool ControllerRenderingEngine::isRunning() const {
+    return m_pThread && m_pThread->isRunning();
+}
+
+void ControllerRenderingEngine::requestEngineSetup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    m_isValid = false;
+    VERIFY_OR_DEBUG_ASSERT(qmlEngine) {
+        kLogger.critical() << "No QML engine was passed!";
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(QThread::currentThread() != thread()) {
+        kLogger.warning() << "Unable to setup OpenGL rendering context from the same "
+                             "thread as the render object";
+        return;
+    }
+    emit engineSetupRequested(qmlEngine);
+
+    const auto lock = lockMutex(&m_mutex);
+    if (!m_quickWindow) {
+        m_waitCondition.wait(&m_mutex);
+    }
+    if (m_isValid) {
+        m_renderControl->prepareThread(m_pThread.get());
+    }
+}
+
+void ControllerRenderingEngine::requestSendingFrameData(
+        Controller* controller, const QByteArray& frame) {
+    emit sendFrameDataRequested(controller, frame);
+}
+
+void ControllerRenderingEngine::setup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    DEBUG_ASSERT(QThread::currentThread() == thread());
```

Comment:
```suggestion
    DEBUG_ASSERT_QOBJECT_THREAD_AFFINITY(this);
```
Technically produces a warnings because `this->member` is bad style. So I guess introduce a `DEBUG_ASSERT_THIS_QOBJECT_THREAD_AFFINITY()` in the same header that calls `thread()` directly (by assuming its used within a QObject; which could even static_assert if you want).
```suggestion
    DEBUG_ASSERT_THIS_QOBJECT_THREAD_AFFINITY();
```